packages/@aws-cdk/assets/test/fs/test.utils.ts

```diff
@@ -192,4 +194,329 @@ export = {
       },
     }
   },
+
+  shouldExcludeDeep: {
+    'without pattern'(test: Test) {
+      testShouldExcludeDeep(test, [], [], ['foo.txt']);
+      testShouldExcludeDeep(test, [''], [], ['foo.txt']);
+      testShouldExcludeDeep(test, ['# comment'], [], ['foo.txt']);
+
+      test.done();
+    },
+    'basic usage'(test: Test) {
+      testShouldExcludeDeep(test, ['foo.txt'], [
+        'foo.txt',
+        'foo.txt/file',
+        'dir/foo.txt',
+      ], [
+        'bar.txt',
+        'foo',
+        'foo.txt.old',
+      ]);
+
+      test.done();
+    },
+    'contridactory'(test: Test) {
+      testShouldExcludeDeep(test, ['foo.txt', '!foo.txt'], [], ['foo.txt']);
```

Comment:
How about contradictory with the order reversed?

Reply:
Done